stumpy/core.py

```diff
@@ -1926,3 +1926,140 @@ def _idx_to_mp(I, T, m, normalize=True):
     P[I < 0] = np.inf
 
     return P
+
+
+@njit()
+def _get_tiles(m, n_A, n_B, diags, tile_size):
```

Comment:
@alvii147 Maybe I'm missing something but when I do:

```
n_A, n_B = 16, 16
excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
diags = np.arange(excl_zone + 1, n_A - m + 1)
tile_size = 5
m = 6
tiles = _get_tiles(m, n_A, n_B, diags, tile_size)
print(tiles)
```

I see:

```
[[[ 0  0]
  [ 5  5]
  [ 3  5]
  [ 3  1]]

 [[ 0  5]
  [ 5  5]
  [-2  5]
  [22  1]]

 [[ 0 10]
  [ 5  1]
  [-4  1]
  [ 5  1]]

 [[ 5  0]
  [ 5  5]
  [ 8  5]
  [ 3  0]]

 [[ 5  5]
  [ 5  5]
  [ 3  5]
  [ 3  1]]

 [[ 5 10]
  [ 5  1]
  [-2  1]
  [ 2  1]]

 [[10  0]
  [ 1  5]
  [13  5]
  [28  0]]

 [[10  5]
  [ 1  5]
  [ 8  5]
  [ 3  0]]

 [[10 10]
  [ 1  1]
  [ 3  1]
  [ 1  0]]]
```

If I understand correctly, when `tile_size=5`, that means that each tile has, at most, a `5 x 5` shape. However, the seventh tile:

```
[[10  0]
  [ 1  5]
  [13  5]
  [28  0]]
```

Is showing a weight of `28` while the maximum weight of any tile should be, I think, `25`. I get that the validity is zero and nothing is computed for this tile but it just feels wrong that the weight can return `28`. Additionally, why not call it `n_dist` instead of `weight`? The latter seems vague. 

Overall, this function looks quite complex. I am guessing that there is a way to refactor this. I'm going to spend some time to investigate it. As you said, since we only call this function once, we don't need this to be super performant.

Reply:
@alvii147 I saw your response in the main PR but it might be better to move the response here. Please :)

Reply:
Whoops my bad, here it is:

@seanlaw thanks for bringing this up, looks like you helped me discover a major bug... 😰 

So earlier, the way I was calculating `tile_weight`, the number of distances to be calculated, was iterating over all our `diags` and literally counting the number of distances we would compute.

Very recently I thought this might be inefficient (which it is) so I changed it to a more dynamic calculation that requires no for loop. But I made a couple of mistake there:
- I forgot to consider the case where the tile is completely invalid (i.e. has zero distances to compute)
- I forgot to take into account tiles that are smaller than `tile_size x tile_size`, i.e. the tiles near the edges.

The case where the tile is completely invalid is what you've pointed out. You're right in that `tile_weight` should never be higher than 25, and should be 0 where `tile_validity` is 0. This one can actually be easily taken care of.

The case where the tile is in the corner/edge of the matrix, that one's a little hard. In your example above, the sixth tile is also wrong:
```
[[ 5 10]
[ 5  1]
[-2  1]
[ 2  1]]
```

If you draw the matrix out, it should show you that this tile's weight should be 3, not 2.

My bad for the buggy code, I'm gonna refactor the code and try to cover all these cases (I think test-driven development might be a good choice here).

TL;DR: I wrote really bad code, and I'm on my way to fixing it!

Thanks a lot for looking into the code though!

> Additionally, why not call it `n_dist` instead of `weight`? The latter seems vague.

I'm just really bad at variable naming, I agree, `n_dist` sounds much more appropriate. Do you have any alternative naming suggestions for `tile_validity` as well? That also sounds a bit weird to me.

Reply:
To count the number of distances, you might as well just use `np.sum(core._count_diagonal_ndist(diags, m, n_A, n_B))`. In your case, it might look more like:

```
n_dist = np.sum(core._count_diagonal_ndist(np.arange(tile_lower_diag, tile_upper_diag), m, tile_height, tile_width))
```

Reply:
> Don't overthink it as this isn't really where your computational time is being spent. Your goal is to simplify the code for easier maintainance.

Yea that's a good point. Looking at `core._count_diagonal_ndist`, it's pretty similar to the naive implementation I had before. I might as well use this for now since this function already exists. Then maybe in the near future we can talk about potentially optimizing `core._count_diagonal_ndist`.

> Additionally, is there really point in assessing/returning any tile that is invalid??

Yepp you're right, there actually isn't any reason to do this.

> I was thinking that we'd either return back a few separate arrays (`tile_offset`, `tile_size`, `tile_ndist`)

Yepp again, I agree, this makes more sense.

Thanks for your feedback again, I'm on it!

Reply:
Done!